### ISSUE TITLE
How to I link to another post from a post in my 11ty blog?

### ISSUE THREAD

**Issue by verlok** [11ty/eleventy#1207]
Hey there,
congrats for creating 11ty, it's awesome.

I'm migrating my blog from Jekyll to 11ty and I'm getting the error

```
`ParseError` was thrown
> tag post_url not found
```

This is because, to cross link my posts, I was doing

```
Ahoy, front-end pirates! Welcome to the Chapter 1 of the 
[Making of The Secret of Front-end Island]
({% post_url 2012-12-09-making-of-the-treasure-of-front-end-island-chapter-1-the-splashing-title %}) 
saga: **the Splashing Title**.
```

That `post_url` is the thing I'm using on Jekyll to crosslink my posts.
Ho can I do that with 11ty?

Thanks

**Comment by mandaputtra** [comment]
You should provide the data of `tag_post`, in [frontmatter](https://www.11ty.dev/docs/data-frontmatter/)

Can you tell me what is cross-link? It is affecting your meta tag? or what?

**Comment by charliepark** [comment]
Is the `2012-12-09-making-of…` bit the URL you want them to get to? As in, you want the reader — if they click on "Making of The Secret of Front-end Island" — to go to the address `http://yoursite.com/2012-12-09-making-of-the-treasure-of-front-end-island-chapter-1-the-splashing-title`?

If that's the case, you don't need to include the Liquid tags, and it'd be regular Markdown syntax for links. You'd have `[Making of The Secret of Front-end Island](2012-12-09-making-of-the-treasure-of-front-end-island-chapter-1-the-splashing-title)`

If I'm misunderstanding, you might find this to be a helpful resource: https://rphunt.github.io/eleventy-walkthrough/template-files.html#linkingcollectionitem

**Comment by verlok** [comment]
For cross link I mean linking from on blog post to another. 

And the thing after { post_url } is the name of the markdown file containing the post. Its not a regular link, because Jekyll removes the date in the generated post url. It’s just an ID, I guess, which is determined by the name of the file.

**Comment by mandaputtra** [comment]
Could you show me some HTML on your blogpost? That cross linking another post?

**Comment by verlok** [comment]
Thank you guys, much appreciated.
I'll keep using Jekyll for the moment, and I'll try 11ty on a new project I'm starting.
Thanks again

**Comment by heatherbooker** [comment]
For me, @charliepark 's comment was almost correct, but if I do:
```
[Making of The Secret of Front-end Island](2012-12-09-making-of-the-treasure-of-front-end-island-chapter-1-the-splashing-title)
```
then I end up at
```
mysite.com/posts/blog-post-you-clicked-from/2012-12-09-making-of-the-treasure...
```
so it doesn't quite work.
If I add a `/` before the url, then I end up at
```
mysite.com/2012-12-09-making-of...
```
so it's missing the `/posts/` part of the url. I settled for manually adding it. If you, dear reader, have a better solution, please I am all ears! :)

TL;DR:

To replace my jekyll links that looked like 
```
[Click here]({{ site.url }}{% post_url 2012-12-09-making-whatever %})
```
I used
```
[Click here](/posts/2012-12-09-making-whatever)
```

Hope that helps someone!

**Comment by pdehaan** [comment]
@heatherbooker what about this?
```md
[Click here](../2012-12-09-making-whatever)
```

sorry, I’m on my phone so I can’t verify at the moment, but since the posts are both in the same parent folder I think it should work (and not require hard coding folder names).

**Comment by nhoizey** [comment]
@verlok did you try this plugin: https://www.npmjs.com/package/eleventy-plugin-link_to

**Comment by RichardJECooke** [comment]
> You should provide the data of `tag_post`, in [frontmatter](https://www.11ty.dev/docs/data-frontmatter/) or another [way](https://www.11ty.dev/docs/data/)


`tag_post` doesn't exist in https://www.11ty.dev/docs/data-frontmatter/. I don't know what you mean by this comment

**Comment by RichardJECooke** [comment]
Why is this issue closed? There's still no way to use a tag like `{% post_link `, as you can in Hexo or Jekyll

**Comment by nhoizey** [comment]
> There's still no way to use a tag like `{% post_link `

@RichardJECooke what about `{% link_to … %}` like I suggested in https://github.com/11ty/eleventy/issues/1207#issuecomment-761692756 ?

**Comment by zachleat** [comment]
@RichardJECooke: The issue was closed because OP closed it. Feel free to file another request for this one!

**Comment by antgel** [comment]
@RichardJECooke I'm in the same boat migrating from Jekyll. The `link_to` plugin worked really smoothly for me, just make sure to enclose the slug in quotes which wasn't necessary with `post_url`.

**Comment by mendhak** [comment]
I'm migrating from Jekyll as well and there's also the [`| url` filter](https://www.11ty.dev/docs/filters/url/) that can help: 

```
[Link to another post]({{ '/second-post' | url }})
```

(This is the first result in search, so wanted to share)

**Comment by pdehaan** [comment]
Here's another rough idea:

### .eleventy.js

```js
/**
 * @param {import("@11ty/eleventy/src/UserConfig")} eleventyConfig
 * @returns {ReturnType<import("@11ty/eleventy/src/defaultConfig")>}
 */
module.exports = function (eleventyConfig) {
  const DOMAIN = "https://localhost:8080";

  eleventyConfig.addFilter("abs_url", function (path="", base=DOMAIN) {
    return new URL(path, base).href;
  });

  // Defaults for Nunjucks templates. LiquidJS will have to explicitly pass `from` and `coll[]`.
  eleventyConfig.addFilter("rel_page", function (to="..", from=this.ctx?.page.url, coll=this.ctx?.collections.all) {
    const absUrl = eleventyConfig.getFilter("abs_url");
    const relUrl = new URL(to, absUrl(from)).pathname;
    const relPage = coll.find(p => p.url === relUrl);
    if (relPage) {
      const { collections, eleventy, pkg, page, ...rest } = relPage.data;
      return {
        ...rest,
        url: page.url,
        page
      };
    }
    throw new Error(`Cannot find "${to}" from "${from}"`);
  });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

### src/blog/2022/11/index.njk

```njk
---
title: Blog/2022/11
---

{%- set Home = "/" | rel_page -%}
{%- set Blog = "/blog/" | rel_page -%}
{%- set Parent = ".." | rel_page -%}

page.url={{ page.url }}
page.url|abs_url={{ page.url | abs_url }}

<nav>
  <menu>
    <li><a href="{{ Home.page.url }}">{{ Home.title }}</a></li>
    <li><a href="{{ Blog.page.url }}">{{ Blog.title }}</a></li>
    <li><a href="{{ Parent.page.url }}">{{ Parent.title }}</a></li>
    <li>{{ title }}</li>
  </menu>
</nav>
```

### src/blog/2022/11/foo.liquid

```liquid
---
title: Blog/2022/11/foo
---

{%- assign Home = "/" | rel_page: page.url, collections.all -%}
{%- assign Blog = "/blog/" | rel_page: page.url, collections.all -%}
{%- assign ParentParent = "../.." | rel_page: page.url, collections.all -%}
{%- assign Parent = ".." | rel_page: page.url, collections.all -%}

{{ page.url }}
{{ page.url | abs_url }}

<nav>
  <menu>
    <li><a href="{{ Home.page.url }}">{{ Home.title }}</a></li>
    <li><a href="{{ Blog.page.url }}">{{ Blog.title }}</a></li>
    <li><a href="{{ ParentParent.page.url }}">{{ ParentParent.title }}</a></li>
    <li><a href="{{ Parent.page.url }}">{{ Parent.title }}</a></li>
    <li>{{ title }}</li>
  </menu>
</nav>
```


### OUTPUT
#### www/blog/2022/11/index.html

```html
page.url=/blog/2022/11/
page.url|abs_url=https://localhost:8080/blog/2022/11/

<nav>
  <menu>
    <li><a href="/">Home</a></li>
    <li><a href="/blog/">Blog</a></li>
    <li><a href="/blog/2022/">Blog/2022</a></li>
    <li>Blog/2022/11</li>
  </menu>
</nav>
```

---

And if you pass an invalid `to` path in the `rel_page` filter, you should get a fatal build error:

```
{%- set Blog = "/blog404/" | rel_page -%}
```

```
Error: Cannot find "/blog2/" from "/blog/2022/11/bar/"
```

**Comment by pdehaan** [comment]
Actually, here's a version that will work w/ Nunjucks and LiquidJS without needing to pass explicit `page.url` and `collections.all` arguments:

```js
eleventyConfig.addFilter("rel_page", function (to="..", from, coll) {
  // `this.ctx` for Nunjucks and `this.context.environments` for LiquidJS.
  const ctx = this.ctx || this.context?.environments;
  if (!ctx) {
    throw new Error("Unable to determine context");
  }
  from = from || ctx.page?.url;
  coll = coll || ctx.collections?.all;
  const absUrl = eleventyConfig.getFilter("abs_url");
  const relUrl = new URL(to, absUrl(from)).pathname;
  const relPage = coll.find(p => p.url === relUrl);
  if (relPage) {
    const { collections, eleventy, pkg, page, ...rest } = relPage.data;
    return {
      ...rest,
      url: page.url,
      page,
    };
  }
  throw new Error(`Cannot find "${to}" from "${from}"`);
});
```

**Comment by mendhak** [comment]
Another idea to throw in, override the Markdown Library's link rendering function, adjust the URL in there: 


```javascript

  const path = require("path");
  const pathPrefix = process.env.PATH_PREFIX || "/my-blog-prefix/";  // Or just / 

  var defaultLinkRender = markdownLibrary.renderer.rules.link_open || function(tokens, idx, options, env, self) {
    return self.renderToken(tokens, idx, options);
  };
  markdownLibrary.renderer.rules.link_open = function (tokens, idx, options, env, self) {

    let link = tokens[0].attrGet('href');
    if(link && link.startsWith("/") && !link.startsWith(pathPrefix)){
      tokens[0].attrSet('href', path.join(pathPrefix, link));
    }
    return defaultLinkRender(tokens, idx, options, env, self);
  };
```

Now in a blog post just write a normal Markdown link:  

    [Link to another page](/this-is-my-first-post/)

I think this can serve basic use cases but will of course break if the posts move, pdehan's solution looks resilient.

**Comment by mendhak** [comment]
This allows linking straight to the Markdown file.  The renderer code will detect it and find the equivalent output URL and replace it.  I like this because it is friendly in the local editor too. 

Link like this:

```markdown
[Link using Markdown path](./customary-lorem-ipsum.md)
```

The code:

```javascript
markdownLibrary.renderer.rules.link_open = function (tokens, idx, options, env, self) {

  let link = tokens[0].attrGet('href');
  if (link && link.endsWith(".md")) {
    if (env.collections?.posts?.[0]) {
      // Match on the file name without extension, eg 'customary-lorem-ipsum', then use its URL in the rendered link.
      let found = env.collections.posts.find(p =>  path.parse(p.inputPath).name === path.parse(link).name);
      tokens[0].attrSet('href', path.join(pathPrefix, found.url));
    }
  }
  return self.renderToken(tokens, idx, options);
}
```

Output:

```html
<a href="/eleventy-mendhak-blog-theme/customary-lorem-ipsum/">Link using Markdown path</a>
```

**Comment by chee** [comment]
a nice way to do this in the upcoming eleventy 3 is to make the link to the file itself and pass it to [the `inputPathToUrl` filter](https://www.11ty.dev/docs/filters/inputpath-to-url/) 

eleventy 3 is in alpha at the time of posting, but this might be helpful for people living on the edge or in the future
